### PR TITLE
Fb pr gating

### PR DESCRIPTION
**Pull request by Jfhelin**
Pipeline and script to handle Gated PRs using Advanced Security for ADO. Currently (October 2023) this is not supported out of the box by the product. We want to restrict new code going into main and only allow PRs if the new code does not introduce any new CodeQL issues. The same could be done for Dependencies with some tweaks.

The idea is to set a branch protection policy (for main), forcing this pipeline to succeed before a PR into main can happen. The pipeline will run a CodeQL scan on the source branch of the PR. Later, using a PowerShell script, the CodeQL issues of the PR source and target will be compared. If there are issues in the PR source that are not in the PR target this pipeline will fail. If new alerts are detected, these will have to be analysed using the regular Advanced Security UI for Code Scanning alerts. Set the branch filter to the new PR branch and fix or dismiss the new alerts. After that, the CIVerify Check for the PR can be re-run, hopefully this time with no issues.

### CODE REVIEW

File: src/pr-gating/CIVerify.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+steps:
+# Run CodeQL on this branch (source branch of the PR) 
+- task: AdvancedSecurity-Codeql-Init@1
+  inputs:
+    languages: 'javascript'
+    querysuite: 'code-scanning'
+
+- task: AdvancedSecurity-Codeql-Autobuild@1
+- task: AdvancedSecurity-Codeql-Analyze@1
+- task: AdvancedSecurity-Publish@1
```
Comment (felickz) [inline src/pr-gating/CIVerify.yml:36]: This task is no longer needed

```suggestion
```

File: src/pr-gating/CIVerify.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+variables:
+  SourcesFolder: src\main
+   
+
+steps:
+# Run CodeQL on this branch (source branch of the PR) 
+- task: AdvancedSecurity-Codeql-Init@1
+  inputs:
+    languages: 'javascript'
+    querysuite: 'code-scanning'
```
Comment (felickz) [inline src/pr-gating/CIVerify.yml:32]: Removing to simplify sample

```suggestion
```

File: src/pr-gating/Setup.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+- Setup [build verification for your main branch](https://learn.microsoft.com/en-us/azure/devops/repos/git/branch-policies?view=azure-devops&tabs=browser#build-validation). Pick the CIVerify pipeline, you can keep the rest of the default settings.
+
+<img width="800" alt="buildpolicy" src="https://github.com/microsoft/GHAzDO-Resources/assets/106392052/74801e80-46e1-4d05-97b1-5f11396330e1">
+
+
+- Figure out a way to test your new setup without adding any bad code to your main branch :) 
+
+## Possible customizations
+- Currently both language (javascript) and query suites (security-extended) are hard coded in CIVerify.yml. These settings might have to be updated to match your needs.
```
Comment (felickz) [inline src/pr-gating/Setup.md:36]:
```suggestion
- The CodeQL language (javascript) is hardcoded in CIVerify.yml. GHAzDO settings will need to be updated to match your needs. 
```

File: src/pr-gating/Setup.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+- Setup [build verification for your main branch](https://learn.microsoft.com/en-us/azure/devops/repos/git/branch-policies?view=azure-devops&tabs=browser#build-validation). Pick the CIVerify pipeline, you can keep the rest of the default settings.
+
+<img width="800" alt="buildpolicy" src="https://github.com/microsoft/GHAzDO-Resources/assets/106392052/74801e80-46e1-4d05-97b1-5f11396330e1">
+
+
+- Figure out a way to test your new setup without adding any bad code to your main branch :) 
+
+## Possible customizations
+- Currently both language (javascript) and query suites (security-extended) are hard coded in CIVerify.yml. These settings might have to be updated to match your needs. 
+- The CI Verification could be extended to also (or only) check new Dependency alerts. The same strategy could be used but tweaks would have to be done to both CIVerify.yml and CIGate.ps1.
```
Comment (felickz) [inline src/pr-gating/Setup.md:37]: - [ ] TODO - validate this is possible

File: src/pr-gating/Setup.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+
+- Figure out a way to test your new setup without adding any bad code to your main branch :) 
+
+## Possible customizations
+- Currently both language (javascript) and query suites (security-extended) are hard coded in CIVerify.yml. These settings might have to be updated to match your needs. 
+- The CI Verification could be extended to also (or only) check new Dependency alerts. The same strategy could be used but tweaks would have to be done to both CIVerify.yml and CIGate.ps1. 
+One complication to concider if you want to use the same strategy for dependency scanning. It could be that the reason we see new issues reported is that the advisory database has been extended since the last scan of main. That is, the new additions in the PR is not the issue but rather that this scan is the first one after an extention of the known vunarabilities. Still, since the alerts can be dismissed in the PR branch, or a new scan can be run on the main branch, I think this is acceptable. 
+- Currently it is hard coded that we compaire to the main branch. If you have a different merge strategy, CIVerify.yml and CIGate.ps1 can be updated to accommodate that.
```
Comment (felickz) [inline src/pr-gating/Setup.md:39]: specifying params for the `base`/`head` branches would be  💯  - default them to the default branch and the current merge branch?

File: src/pr-gating/Setup.md (hunk 4 above)
Comment (Jfhelin) [reply src/pr-gating/Setup.md:37]: Checked and yes. They will not show up in the branch filter for dependencies. Still, they are accessible via the API and same technique can be used.

File: src/pr-gating/Setup.md (hunk 5 above)
Comment (Jfhelin) [reply src/pr-gating/Setup.md]: This was an easy change. I also updated the documentation to make it clear that this check is generic.

File: src/pr-gating/CIGate.ps1 [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+$repositoryId = ${env:BUILD_REPOSITORY_ID} 
+$prTargetBranch = ${env:SYSTEM_PULLREQUEST_TARGETBRANCH}
+$prSourceBranch = ${env:BUILD_SOURCEBRANCH}
+$pair = ":${pass}"
+$bytes = [System.Text.Encoding]::ASCII.GetBytes($pair)
+$base64 = [System.Convert]::ToBase64String($bytes)
+$basicAuthValue = "Basic $base64"
+$headers = @{ Authorization = $basicAuthValue }
+
+$urlTargetAlerts = "https://advsec.dev.azure.com/{0}/{1}/_apis/Alert/Repositories/{2}/Alerts?top=5000&orderBy=lastSeen&criteria.alertType=3&criteria.branchName={3}&criteria.onlyDefaultBranchAlerts=true&useDatabaseProvider=true" -f $orgName, $project, $repositoryId, $prTargetBranch
```
Comment (felickz) [inline src/pr-gating/CIGate.ps1:16]: This script appeared broken for me ... 

I updated these in the commit: [Fix Alert URLs](https://github.com/microsoft/GHAzDO-Resources/pull/14/commits/d166721eadf2f6a5e590e4100a7819bd55b475f6)

File: src/pr-gating/Setup.md (hunk 4 above)
Comment (felickz) [reply src/pr-gating/Setup.md:38]: Confirmed the same:

```
S C:\repos\microsoft\GHAzDO-Resources> . 'C:\repos\microsoft\GHAzDO-Resources\src\pr-gating\CIGate.ps1'
Will check to see if there are any new  issues in this PR branch
PR source : refs/pull/1/merge. PR target: refs/heads/master
##[error] There are more alerts in the PR source branch refs/pull/1/merge, compared to target branch refs/heads/master :

##vso[task.logissue type=error;sourcepath=WebGoat.NET/Views/Shared/_LoginPartial.cshtml WebGoat.NET/Views/Shared/_LoginPartial.cshtml;linenumber=20 20;columnnumber=113 113] New  issue detected #28 : Cross-site scripting (cs/web/xss).
##[error] Fix or dismiss this new alert in the Advanced Security UI for pr branch refs/pull/1/merge.
##[error] Details for this new alert:  https://dev.azure.com/octodemo-felickz/_git/proj-felickz/alerts/28?branch=refs/pull/1/merge

##vso[task.logissue type=error;sourcepath=s/WebGoat.NET/WebGoat.NET.csproj;linenumber=;columnnumber=] New  issue detected #152 : Cookie parsing failure (GHSA-hxrm-9w7p-39cc).
##[error] Fix or dismiss this new alert in the Advanced Security UI for pr branch refs/pull/1/merge.
##[error] Details for this new alert:  https://dev.azure.com/octodemo-felickz/_git/proj-felickz/alerts/152?branch=refs/pull/1/merge

##[error] Please review these Code Scanning alerts for the  branch using the regular Advanced Security UI
##[error] Dissmiss or fix the issues listed and try re-queue the CIVerify task.
```

Comment (felickz) [reply src/pr-gating/Setup.md:38]: Added 
[Specify no alert type (code and dependency alerts)](https://github.com/microsoft/GHAzDO-Resources/pull/14/commits/ecfefc7104b8f83d1162d6394a4bd6aa5931a465)

**Comment by felickz** [comment]
Fixed bug in the link to the alert: 
[Fix alert link to include repository](https://github.com/microsoft/GHAzDO-Resources/pull/14/commits/78523d51802936c9e470f463b330fc26cf71ba4c)

File: src/pr-gating/CIGate.ps1 (hunk 6 above)
Comment (Jfhelin) [reply src/pr-gating/CIGate.ps1]: Yes, correct. I messed up in copying... Thanks for correcting.

**Review by ncouraud — APPROVED** [review]

**Pull request merged**